### PR TITLE
[CARBONDATA-4014] Support Change Column Comment

### PR DESCRIPTION
**Pull request by marchpure**
### Why is this PR needed?
 Now, we support add column comment in "CREATE TABLE" and "ADD COLUMN". but do not support alter comment of the column, which shall be support in 'CHANGE COLUMN'
 
 ### What changes were proposed in this PR?
Support "ALTER TABLE table_name CHANGE [COLUMN] col_name col_name data_type [COMMENT col_comment]"
    
 ### Does this PR introduce any user interface change?
 - Yes. add CHANGE COLUMN explaination in the ddl document.

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4228/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2484/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4229/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2485/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4230/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2486/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4232/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2488/

**Comment by Klaus-xjp** [comment]
LGTM

File: integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala [hunk 1]
```diff
@@ -366,7 +366,7 @@ class AlterTableValidationTestCase extends QueryTest with BeforeAndAfterAll {
     sql("alter table default.restructure change decimalfield deciMalfield Decimal(11,3)")
     sql("alter table default.restructure change decimalfield deciMalfield Decimal(12,3)")
     intercept[ProcessMetaDataException] {
-      sql("alter table default.restructure change decimalfield deciMalfield Decimal(12,3)")
+      sql("alter table default.restructure change decimalfield deciMalfield Decimal(12,2)")
```
Comment (Zhangshunyu) [inline integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala:369]: Why change here?

Comment (marchpure) [reply integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala:369]: Before this pr, we don't allow change dataype to same precision, which is actually allowed now for 2 reason:
1) Be consistent with hive and spark;
2) In 'CHANGE COLUMN' command. we support just change comment without renaming and changing datatype;

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 2]
```diff
@@ -169,7 +169,8 @@ case class AlterTableDataTypeChangeModel(dataTypeInfo: DataTypeInfo,
     tableName: String,
     columnName: String,
     newColumnName: String,
-    isColumnRename: Boolean)
+    isColumnRename: Boolean,
+    newColumnComment: Option[String])
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala:173]: better to provide a default value:  None.
it can avoid other unnecessary changes.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala [hunk 3]
```diff
@@ -150,45 +140,45 @@ private[sql] case class CarbonAlterTableColRenameDataTypeChangeCommand(
...
+      // if column datatype change operation is on partition column, then fail the
+      // chang column operation
+      if (null != carbonTable.getPartitionInfo) {
+        val partitionColumns = carbonTable.getPartitionInfo.getColumnSchemaList
+        partitionColumns.asScala.foreach {
+          col =>
+            if (col.getColumnName.equalsIgnoreCase(oldColumnName)) {
+              throw new MalformedCarbonCommandException(
+                s"Column Rename Operation failed. Renaming or ChangeType or " +
+                  s"ChangeComment the partition column $newColumnName is not " +
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala:171]: how about to support changing comments of partition columns?

File: integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala [hunk 4]
```diff
@@ -1135,11 +1135,7 @@ object CarbonParserUtil {
         }
         DataTypeInfo("decimal", precision, scale)
       case _ =>
-        if (isColumnRename) {
-          DataTypeInfo(DataTypeConverterUtil.convertToCarbonType(dataType).getName.toLowerCase)
-        } else {
-          throw new MalformedCarbonCommandException("Data type provided is invalid.")
-        }
+        DataTypeInfo(DataTypeConverterUtil.convertToCarbonType(dataType).getName.toLowerCase)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala:1138]: is the parameter isColumnRename of this method needed?

**Review by QiangCai — COMMENTED** [review]
1. do we support alter column without any change?
2. how to remove comment from column

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/DDLHelper.scala [hunk 5]
```diff
@@ -230,14 +231,22 @@ object DDLHelper {
           .toLowerCase,
         values,
         isColumnRename)
+      var newColumnComment: Option[String] = Option.empty
+      if (changeColumnCommand.newColumn.metadata != null &&
+        changeColumnCommand.newColumn.metadata.contains(CarbonCommonConstants.COLUMN_COMMENT)) {
+        newColumnComment = Some(changeColumnCommand.
+          newColumn.metadata.getString(CarbonCommonConstants.COLUMN_COMMENT))
+      }
```
Comment (Kejian-Li) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/DDLHelper.scala:239]: there is already a varibale "newColumn" that can be used above,  or it is better to make "changeColumnCommand.newColumn.metadata" as a variable.

File: integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala (hunk 4 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/catalyst/CarbonParserUtil.scala:1137]: Yes. useless to throw MalformedCarbonCommandException("Data type provided is invalid.")
CarbonAlterTableColumnRenameCommand.validateColumnDataType will validate the given datatype

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala (hunk 3 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala]: It is really hard to support changing comments of partition columns, as spark and hive also don't support.
https://stackoverflow.com/questions/57871294/how-to-add-edit-the-comment-of-a-partitioned-column-in-hive-table

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala (hunk 2 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: I have modified code according to you suggestion.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/DDLHelper.scala (hunk 5 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/DDLHelper.scala]: I have modified code according to you suggestion.

**Comment by marchpure** [comment]
> 1. do we support alter column without any change?
> 2. how to remove comment from column

1. now, we support alter column without any change.
2. now. we don't support remove comment.

**Comment by QiangCai** [comment]
if alter column change nothing, the command should return directly without execution.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4252/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2507/

**Comment by marchpure** [comment]
> if alter column change nothing, the command should return directly without execution.

I have modified code according to you suggestion.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2517/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4262/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala [hunk 6]
```diff
@@ -258,7 +262,7 @@ private[sql] case class CarbonAlterTableColRenameDataTypeChangeCommand(
       }
     } catch {
       case e: Exception =>
-        if (carbonTable != null) {
+       if ( carbonTable != null) {
```
Comment (Klaus-xjp) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala:265]: there has a blank between the '(' and carbonTable

Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala:265]: I have modified code according to you suggestion.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2520/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4265/

**Comment by marchpure** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4266/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2521/

**Comment by Zhangshunyu** [comment]
LGTM

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala [hunk 7]
```diff
@@ -150,45 +141,50 @@ private[sql] case class CarbonAlterTableColRenameDataTypeChangeCommand(
...
+      }
+      // if column datatype change operation is on partition column, then fail the
+      // chang column operation
+      if (null != carbonTable.getPartitionInfo) {
+        val partitionColumns = carbonTable.getPartitionInfo.getColumnSchemaList
+        partitionColumns.asScala.foreach {
+          col =>
+            if (col.getColumnName.equalsIgnoreCase(oldColumnName)) {
+              throw new MalformedCarbonCommandException(
+                s"Change Column Operation failed. Renaming or ChangeType or " +
```
Comment (kunal642) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/schema/CarbonAlterTableColRenameDataTypeChangeCommand.scala:176]: Please change this to 

`throw new InvalidOperationException("Alter on partition columns is not supported")`

And add a test case for this

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4268/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2523/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4269/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2524/

**Comment by kunal642** [comment]
LGTM
